### PR TITLE
Update misc code for RISC-V in OpenJ9 (part7/misc)

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes mainly include the remaining
miscellaneous code in various components
to enable RISC-V 64bit from the OpenJ9 perspective.

Issue: eclipse#7421

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Already verified the compilation on Fedora_Linux/QEMU.

Reviewer: @pshipton , @DanHeidinga

File: runtime/vm/jvminit.c [hunk 1]
```diff
@@ -2771,6 +2771,12 @@ modifyDllLoadTable(J9JavaVM * vm, J9Pool* loadTable, J9VMInitArgs* j9vm_args)
 			}
 		}
 	}
+	
+	/* Temporarily disable JIT/AOT until it is fully implemented */
+#if defined(RISCV64)
+	xint = TRUE;
+	xjit = xaot = xnoaot = FALSE;
```
Comment (keithc-ca) [inline runtime/vm/jvminit.c]: One assignment at a time, please.

Comment (ChengJin01) [reply runtime/vm/jvminit.c]: Fixed.

File: runtime/port/unix/j9sysinfo.c [hunk 2]
```diff
@@ -1187,6 +1193,22 @@ getS390Description(struct J9PortLibrary *portLibrary, J9ProcessorDesc *desc)
...
+ * Populates J9ProcessorDesc *desc on RISC-V
+ *
+ * @param[in] desc pointer to the struct that will contain the CPU type and features.
+ *
+ * @return 0 on success, -1 on failure
+ */
+static intptr_t
+getRISCV64Description(struct J9PortLibrary *portLibrary, J9ProcessorDesc *desc)
+{
+	return 0;
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/port/unix/j9sysinfo.c]: Agreed and the code was added in there against the implementation of `omrsysinfo_get_riscv_description  `at `/omr/port/unix/omrsysinfo.c`.

File: runtime/port/unix/j9sysinfo.c [hunk 3]
```diff
@@ -1628,6 +1650,21 @@ j9sysinfo_get_cache_info(struct J9PortLibrary *portLibrary, const J9CacheInfoQue
...
+	 * In such case, there is no padding adjustment for class objects
+	 * even if the -XX:-RestrictContended option is specified on the command line.
+	 */
+#if defined(RISCV64)
+	DIR* cacheDir = opendir("/sys/devices/system/cpu/cpu0/cache");
+	if (ENOENT == errno) {
+		Trc_PRT_sysinfo_get_cache_info_exit(result);
+		return result;
+	}
+#endif
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/port/unix/j9sysinfo.c]: Agreed and fixed against the suggestion above.

File: runtime/tests/port/si.c [hunk 4]
```diff
@@ -2183,7 +2183,7 @@ j9sysinfo_test_get_l1dcache_line_size(struct J9PortLibrary *portLibrary)
 	cQuery.level = 1;
 	cQuery.cacheType = J9PORT_CACHEINFO_DCACHE;
 	rc = j9sysinfo_get_cache_info(&cQuery);
-#if !(defined(J9ARM))
+#if !(defined(J9ARM) || defined(RISCV64))
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/tests/port/si.c]: The comments were updated.

File: runtime/tests/port/si.c [hunk 5]
```diff
@@ -2273,7 +2273,7 @@ j9sysinfo_runTests(struct J9PortLibrary *portLibrary, char *argv0)
 	rc |= j9sysinfo_test_get_groups(portLibrary);
 #endif /* !(defined(WIN32) || defined(WIN64)) */
 	rc |= j9sysinfo_test_get_l1dcache_line_size(portLibrary);
-#if !(defined(LINUXPPC) || defined(S390) || defined(J9ZOS390) || defined(J9ARM) || defined(OSX))
+#if !(defined(LINUXPPC) || defined(S390) || defined(J9ZOS390) || defined(J9ARM) || defined(OSX) || defined(RISCV64))
 	rc |= j9sysinfo_test_get_levels_and_types(portLibrary);
 #endif /* !(defined(LINUXPPC) || defined(S390) || defined(J9ZOS390) || defined(J9ARM) || defined(OSX)) */
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/tests/port/si.c]: The comment was updated.

File: runtime/port/unix/j9sysinfo.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply runtime/port/unix/j9sysinfo.c]: The failure test isn't right - a successful call to `opendir()` will not modify `errno`: the check should be `NULL != cacheDir`.

[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/port/unix/j9sysinfo.c]: Changed the code as suggested above.

File: runtime/port/unix/j9sysinfo.c [hunk 6]
```diff
@@ -1628,6 +1652,23 @@ j9sysinfo_get_cache_info(struct J9PortLibrary *portLibrary, const J9CacheInfoQue
...
+	 * even if the -XX:-RestrictContended option is specified on the command line.
+	 */
+#if defined(RISCV64)
+	DIR* cacheDir = opendir("/sys/devices/system/cpu/cpu0/cache");
+	if (NULL != cacheDir) {
+		closedir(cacheDir);
+	} else {
+		Trc_PRT_sysinfo_get_cache_info_exit(result);
+		return result;
+	}
```
Comment (keithc-ca) [inline runtime/port/unix/j9sysinfo.c]: This isn't legal C code (the declaration of `cacheDir` follows the call to `Trc_PRT_sysinfo_get_cache_info_enter()` above): nest these 7 lines in a block to fix that.
The comment belongs within the `#if defined(RISCV64)` section as well.

**Comment by DanHeidinga** [comment]
@ChengJin01 moving the discussion of risc-v out of #8617, it may be worth testing the `sysconf` approach on the risc-v systems (emulated and hw) as a workaround for the missing /sys files in QEMU mode

**Comment by ChengJin01** [comment]
@DanHeidinga, will get started to set things up once the work on MH.Lookup (dropLookupMode() & test cases specific to MH.Lookup) is finished.

**Comment by ChengJin01** [comment]
@DanHeidinga,  as suggested in https://stackoverflow.com/questions/794632/programmatically-get-the-cache-line-size
```
On Linux look at sysconf(3).
sysconf (_SC_LEVEL1_DCACHE_LINESIZE)
You can also get it from the command line using getconf:
$ getconf LEVEL1_DCACHE_LINESIZE
64
```
So I tried `getconf LEVEL1_DCACHE_LINESIZE` on Fedora_RISCV/QEMU and it ended up with zero as follows, which indicates the approach is unsuitable for the Fedora_RISCV via QEMU.
```
[root@stage4 ~]# uname -a
Linux stage4.fedoraproject.org 4.19.0-rc8 
#1 SMP Wed Oct 17 15:11:25 UTC 2018 riscv64 riscv64 riscv64 GNU/Linux

[root@stage4 ~]# getconf LEVEL1_DCACHE_LINESIZE
0
```

**Comment by ChengJin01** [comment]
@keithc-ca,  

Based on the result bove, it might be better to call `sysconf (_SC_LEVEL1_DCACHE_LINESIZE)` as compared to the check on the existence of `/sys/devices/system/cpu/cpu0/cache` by calling `opendir`
e.g. the code might look like
```
#elif defined(LINUX) && defined(_SC_LEVEL1_DCACHE_LINESIZE)
	if (query->cmd == J9PORT_CACHEINFO_QUERY_LINESIZE
	    && query->cacheType == J9PORT_CACHEINFO_DCACHE
	    && query->level == 1) {
		result = (int32_t)sysconf(_SC_LEVEL1_DCACHE_LINESIZE);
		if (result <= 0) { <--------------------------
			result = J9PORT_ERROR_SYSINFO_NOT_SUPPORTED;
		}
	} else {
		result = J9PORT_ERROR_SYSINFO_NOT_SUPPORTED;
	}
#endif
```
So it ends up with J9PORT_ERROR_SYSINFO_NOT_SUPPORTED if the result <= 0

**Comment by keithc-ca** [comment]
> @keithc-ca,
> 
> Based on the result above, it might be better to call `sysconf(_SC_LEVEL1_DCACHE_LINESIZE)`
> 
> So it ends up with `J9PORT_ERROR_SYSINFO_NOT_SUPPORTED`

Does QEMU have a command-line or configuration option that would allow us to specify the cache line size? (I looked quickly and didn't find anything promising).

**Comment by ChengJin01** [comment]
> Does QEMU have a command-line or configuration option that would allow us to specify the cache line size? (I looked quickly and didn't find anything promising).

I went over the document specific to the command line options at https://qemu.weilnetz.de/doc/qemu-doc.html and none of them is related to the cache line size. So it is not supported in QEMU.

In terms of checking the cache line size, there should be no big difference between sysconf (_SC_LEVEL1_DCACHE_LINESIZE) (return 0 if doesn't exist; otherwise let it through) should do petty much the same as opendir (check the specified directory to determine whether the cache line size is ready for use and return if doesn't exist).

**Comment by keithc-ca** [comment]
If neither approach gives us the information we want then, yes, that's no difference - but not a happy situation.

I would prefer using `sysconf()`.

When will we have real hardware so this issue becomes moot?

**Comment by ChengJin01** [comment]
> 
> 
> If neither approach gives us the information we want then, yes, that's no difference - but not a happy situation.
> 
> I would prefer using `sysconf()`.
> 
> When will we have real hardware so this issue becomes moot?

We've already receive the Dev board and will get back to address this once I finished the work left in MH.Lookup.

File: runtime/port/unix/j9sysinfo.c (hunk 6 above)
Comment (ChengJin01) [reply runtime/port/unix/j9sysinfo.c]: @keithc-ca ,  the code there has been fixed against your comment.

In addition, I checked the result cache directory on the HiFive U540 Dev board (booting Fedora_riscv) and the results are as follows:
```
[root@openj9riscv ~]# ls  -l /sys/devices/system/cpu/cpu0/cache/index0
total 0
-r--r--r--. 1 root root 4096 Mar 30 20:02 coherency_line_size
-r--r--r--. 1 root root 4096 Dec 15 05:47 level
-r--r--r--. 1 root root 4096 Mar 30 20:10 number_of_sets
-r--r--r--. 1 root root 4096 Mar 30 20:10 shared_cpu_list
-r--r--r--. 1 root root 4096 Mar 30 20:10 shared_cpu_map
-r--r--r--. 1 root root 4096 Dec 15 05:47 size
-r--r--r--. 1 root root 4096 Dec 15 05:47 type
-rw-r--r--. 1 root root 4096 Mar 30 20:10 uevent
-r--r--r--. 1 root root 4096 Mar 30 20:10 ways_of_associativity

[root@openj9riscv ~]# cat  /sys/devices/system/cpu/cpu0/cache/index0/level
1
[root@openj9riscv ~]# cat  /sys/devices/system/cpu/cpu0/cache/index0/coherency_line_size
64  <------
[root@openj9riscv ~]# cat  /sys/devices/system/cpu/cpu0/cache/index1/level
1
[root@openj9riscv ~]# cat  /sys/devices/system/cpu/cpu0/cache/index1/coherency_line_size
64 <------

[jincheng_riscv@openj9riscv ~]$ getconf LEVEL1_DCACHE_LINESIZE
0 <------ (no data out there)
```
Based on results above, it indicates:
1) these cache related directories do exist when the Linux image is booted on the hardware. 2) there is no support for `sysconf `at least on Fedora_riscv.

After talking with the Fedora/RISCV developer at #fedora-riscv/IRC as to `sysconf `,  I was told this function can be can added to Fedora/RISCV but this is not in their wishlist (not priority to do so) for now and there is no plan when this function is supported on Fedora/RISCV.

Under such circumstance, we have to stick to check the cache related directory as it is more reliable as compared to `sysconf ` on the hardware.

**Comment by keithc-ca** [comment]
This looks good to me. Does this depend on anything that isn't already merged in OMR or OpenJ9?

**Comment by ChengJin01** [comment]
> This looks good to me. Does this depend on anything that isn't already merged in OMR or OpenJ9?

This PR mainly puts together miscellaneous/left things like trace, sysinfo, etc, which should have no direct dependency on the pending PRs as  https://github.com/eclipse/openj9/pull/7442/files (JIT) & https://github.com/eclipse/openj9/pull/7445/files (VM) in terms of headers

**Comment by keithc-ca** [comment]
We have no PR builds for RISC-V yet - the travis build is sufficient validation.

**Pull request merged**